### ISSUE TITLE
TLS client auth to backends

### ISSUE THREAD

**Issue by ishitatsuyuki** [traefik/traefik#3038]
## Proxy mutual TLS configuration proposal

This proposal adds mutual TLS  authentication with backends, as well as some other configuration format improvements.

### Added options

- Support for secrets are added. Secrets should be watched if available for hot reload.
- For all server certificate verification, add support for certification revocation lists and verification depth. See below.
- For specifying client certificate, the following options are used:
  ```toml
  [traefik.<service-name>.proxyCert]
  file = "secretkey" # file, content or secret
  serverName = "somehost" # Use this hostname for SNI and subject name verification
  ```
- For specifying server certificate, the contents of the following `ServerCA` are placed under the key `[traefik.<service-name>.proxyCA]`.

### Enhancement to existing configurations

- `FileOrContent` should also accept secrets as well, if supported in backend.
  * What should be the syntax for this? Something like `kubernetes:secret_name`?
- CA related configurations should be refactored so they have unity in syntax:
  ```toml
  [somepath.ClientCA] # Or ServerCA, renamed from RootCA
  files = ["..."] # file, content or secret
  crl = ["..."] # file, content or secret
  verify = true # renamed/flipped from optional
  verifyDepth = 1 # 0 -> self-signed only, 1 -> one level of CA, and so on...
  ```
  * Deprecate `RootCAs` and `InsecureSkipVerify`.
  * This change may create confusion for old users.

All the namings are for reference. Feel free to suggest anything better (less confusing).

[nginx doc](https://nginx.org/en/docs/http/ngx_http_proxy_module.html) is used as a strong reference on designing this.

**Comment by ivoribeiro** [comment]
+1

**Comment by borgesp** [comment]
+1

**Comment by ruiaraujo1** [comment]
+1

**Comment by traefiker** [comment]
Hi! I'm Træfiker :robot: the bot in charge of communication regulation.

The best ways to participate in an issue are
- to show your support by [adding a reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the initial message
- to provide additionnal information
- to propose a potential fix and create a Pull Request

These rules preserve the readability of the issue and help us be more efficient.

Thanks for your help.

**Comment by bsukramlitt** [comment]
+1

**Comment by arno01** [comment]
@ishitatsuyuki thank you for creating this proposal.

I would like to add one more point to it:

- Allow Traefik generate self-signed x509 certificates (following the best TLS practices) per service;

This would cover the case when a user either does not know how-to or not willing to bother with x509's bits, but wants Traefik to manage insecure TLS (self-signed x509 bits) between the Traefik and backend services.
(Similar cases to https://github.com/minio/minio/issues/6093 , https://github.com/minio/minio/issues/6125)

**Comment by tscibilia** [comment]
Looking for an update on allowing traefik to route to https backend instead of giving a 500 error. 

I'm trying to get a proxmox webgui to work with traefik but since it only allows a https connections, it's become impossible ([see unanswered stackoverflow topic](https://stackoverflow.com/questions/53396467/file-section-fails-with-internal-server-error-for-proxmox))

**Comment by tomwiggers** [comment]
Is there any progress on this? I would really like to be able to configure Traefik to allow a self signed certificate on services.
I have cockroachdb, but that admin ui uses a self signed cert. Now I cannot use Traefik to access it, but it seems like such a basic and trivial feature.

**Comment by branchmispredictor** [comment]
I'm in strong support of this proposal. One thing I would like to suggest is similarly reworking the [ForwardAuth TLS config](https://docs.traefik.io/middlewares/forwardauth/#tls) or pulling it into this effort as well.

**Comment by jbdoumenjou** [comment]
Closed by #7203
